src/org/opensolaris/opengrok/util/AcceptHelper.java

```diff
@@ -0,0 +1,236 @@
+                (!(file.isDirectory() || env.getIncludedNames().match(file)))) {
+            return false;
+        }
+
+        String absolutePath = file.getAbsolutePath();
+
+        if (env.getIgnoredNames().ignore(file)) {
+            LOGGER.log(Level.FINER, "ignoring {0}", absolutePath);
+            return false;
+        }
+
+        if (!file.canRead()) {
+            LOGGER.log(Level.WARNING, "Could not read {0}", absolutePath);
+            return false;
+        }
+
+        try {
+            String canonicalPath = file.getCanonicalPath();
+            if (!absolutePath.equals(canonicalPath)
+                    && !acceptSymlink(project, file)) {
+
+                LOGGER.log(Level.FINE, "Skipped symlink ''{0}'' -> ''{1}''",
+                        new Object[]{absolutePath, canonicalPath});
+                return false;
+            }
+            //below will only let go files and directories, anything else is considered special and is not added
+            if (!file.isFile() && !file.isDirectory()) {
+                LOGGER.log(Level.WARNING, "Ignored special file {0}",
+                        absolutePath);
+                return false;
+            }
+        } catch (IOException exp) {
+            LOGGER.log(Level.WARNING, "Failed to resolve name: {0}",
+                    absolutePath);
+            LOGGER.log(Level.FINE, "Stack Trace: ", exp);
+        }
+
+        if (file.isDirectory()) {
+            // always acceptFile directories so that their files can be examined
+            return true;
+        }
+
+        if (HistoryGuru.getInstance().hasHistory(file)) {
+            // versioned files should always be accepted
+            return true;
+        }
+
+        // this is an unversioned file, check if it should be indexed
+        return !env.isIndexVersionedFilesOnly();
+    }
+
+    /**
+     * Check if this file should be accepted. This method has the same effect as
+     * the call accept(null, parent, file).
+     *
+     * @see AcceptHelper#accept(Project, File, File)
+     * @param parent parent directory of the file
+     * @param file the file to check
+     * @return true if the file should be included, false otherwise
+     */
+    public static boolean accept(File parent, File file) {
+        return accept(null, parent, file);
+    }
+
+    /**
+     * Check if this file should be accepted.
+     *
+     * @param project if the file is relevant to some project (may be null)
+     * @param parent parent directory of the file
+     * @param file the file to check
+     * @return true if the file should be included, false otherwise
+     */
+    public static boolean accept(Project project, File parent, File file) {
+        try {
+            File f1 = parent.getCanonicalFile();
+            File f2 = file.getCanonicalFile();
+            if (f2.equals(f1)) {
+                LOGGER.log(Level.INFO, "Skipping links to itself...: {0} {1}",
```

Comment:
these messages should be self-explanatory (i.e. make it clear what are the strings)